### PR TITLE
Generalized ES Sink

### PR DESCRIPTION
**Pull request by dconger**
This implementation is modeled after `Ecto.Repo`. To see this implemented, see either the `Rivulet.TestDan` module within `rivulet` or the `SnowdenSink.ElasticSearchSink` module within the `platform` repo's `dan/test-for-staging` branch.

### CODE REVIEW

File: lib/rivulet/elastic_search_sink/config.ex [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+     topic: "platform_nps_location_joins"
+   ]
+  """
+  def from_sink_opts(opts) do
+    callback_module = Keyword.get(opts, :callback_module)
+    elastic_index = Keyword.get(opts, :elastic_index)
+    elastic_url = Keyword.get(opts, :elastic_url, [])
+    elastic_type = Keyword.fetch!(opts, :elastic_type)
+    topic = Keyword.fetch!(opts, :topic)
+    elastic_mapping = Keyword.get(opts, :elastic_mapping)
```
Comment (CJPoll) [inline lib/rivulet/elastic_search_sink/config.ex]: get vs fetch! where appropriate 
fetch! for required fields
get for optional, no default if nil is acceptable, default if nil is unacceptable

File: lib/rivulet/elastic_search_sink/consumer.ex [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    consumer_config =
+      %Config{
+        client_id: Rivulet.client_name(),
+        consumer_group_name: config.consumer_group,
+        topics: [config.topic],
+        group_config: [
+          offset_commit_policy: :commit_to_kafka_v2,
+          offset_commit_interval_seconds: 1,
+        ],
+        consumer_config: [begin_offset: :earliest, max_bytes: 2_000_000]
```
Comment (CJPoll) [inline lib/rivulet/elastic_search_sink/consumer.ex]: `Rivulet.Config.max_bytes`

File: lib/rivulet/elastic_search_sink/consumer.ex [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  @doc """
+  NOTE: the sink_supervisor_process was passed in via the self() portion of the
+  call within supervisor.ex:
+
+  worker(Rivulet.ElasticSearchSink.ConsumerTwo, [config, self()], id: :consumer)
+  """
+  def start_link(%SinkConfig{} = config, sink_supervisor_process) do
+    consumer_config =
+      %Config{
+        client_id: Rivulet.client_name(),
```
Comment (CJPoll) [inline lib/rivulet/elastic_search_sink/consumer.ex]: client_name!

File: lib/rivulet/elastic_search_sink/database/elastic_search_generator.ex [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+  # Errors
+  @spec handle_es_response({:ok | any, %HTTPoison.Response{}}, action_type) :: true | exception
+  defp handle_es_response({_, %HTTPoison.Response{body: %{"error" => %{"reason" => reason}}}}, action) do
+    raise "Elasticsearch error: #{action} #{reason}"
+  end
+
+  # Success
+  defp handle_es_response({:ok, %HTTPoison.Response{status_code: status_code}}, _) when status_code in 200..299 do
+    true
```
Comment (CJPoll) [inline lib/rivulet/elastic_search_sink/database/elastic_search_generator.ex]: `:ok`

File: lib/rivulet/elastic_search_sink/supervisor.ex [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+  end
+
+  def init({caller, otp_app, opts}) do
+    case runtime_config(:supervisor, caller, otp_app, opts) do
+      {:ok, consumer_opts} ->
+        %Config{} = config = Config.from_sink_opts(consumer_opts)
+
+        {:ok, _} = Application.ensure_all_started(:httpoison)
+
+        count = 1
```
Comment (CJPoll) [inline lib/rivulet/elastic_search_sink/supervisor.ex]: Make configurable

File: lib/rivulet/elastic_search_sink/consumer.ex [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    Consumer,
+    Consumer.Config,
+    Kafka.Partition,
+  }
+
+  alias Rivulet.ElasticSearchSink.Config, as: SinkConfig
+
+  defmodule State do
+    @enforce_keys [:sink_consumer_pid]
+    defstruct [:manager_pid, :sink_consumer_pid]
```
Comment (CJPoll) [inline lib/rivulet/elastic_search_sink/consumer.ex]: sink_consumer_pid references a supervisor, not a consumer

File: lib/rivulet/elastic_search_sink/supervisor.ex [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        :ignore ->
+          :ignore
+      end
+    else
+      raise ArgumentError,
+        "configuration for #{inspect repo} not specified in #{inspect otp_app} environment"
+    end
+  end
+
+  defp repo_init(type, repo, config) do
```
Comment (CJPoll) [inline lib/rivulet/elastic_search_sink/supervisor.ex]: Change name

File: lib/rivulet/elastic_search_sink/writer.ex [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+      |> filter_for_successfully_inserted(messages)
+
+    state.callback_module.on_complete(successfully_inserted)
+
+    Rivulet.Consumer.ack(Rivulet.client_name!, partition, offset)
+
+    {:noreply, state}
+  end
+
+  def filter_for_successfully_inserted(%{"errors" => false} = _es_response, records) do
```
Comment (CJPoll) [inline lib/rivulet/elastic_search_sink/writer.ex]: This stuff should be in another module

File: lib/rivulet/join/funcs.ex [hunk 9]
```diff
@@ -12,7 +12,7 @@ defmodule Rivulet.Kafka.Join.Funcs do
           offset_commit_interval_secons: 1,
         ],
         consumer_config: [begin_offset: :earliest, max_bytes: Rivulet.Config.max_bytes()]
-      } |> IO.inspect(label: "join funcs config")
```
Comment (CJPoll) [inline lib/rivulet/join/funcs.ex]: Please print out configs in some fashion

File: lib/rivulet/join/funcs.ex [hunk 10]
```diff
@@ -76,15 +76,19 @@ defmodule Rivulet.Kafka.Join.Funcs do
 
     bulk =
       messages
+      # |> Enum.map(fn(message) -> {UUID.uuid4(), message} end)
```
Comment (CJPoll) [inline lib/rivulet/join/funcs.ex]: Commented code

File: lib/rivulet/kafka/transformer/avro_inspect.ex [hunk 11]
```diff
@@ -4,7 +4,6 @@ defmodule Rivulet.Transformer.AvroInspect do
   def handle_message(%Rivulet.Kafka.Consumer.Message{} = m) do
     with {:ok, key} <- Avro.decode(m.raw_key),
          {:ok, value} <- Avro.decode(m.raw_value) do
-           IO.inspect({key, value})
```
Comment (CJPoll) [inline lib/rivulet/kafka/transformer/avro_inspect.ex]: Let's leave the IO.inspect here

File: lib/rivulet/sql_sink/writer/manager.ex [hunk 12]
```diff
@@ -16,7 +16,7 @@ defmodule Rivulet.SQLSink.Writer.Manager do
     {:ok, {sup, count}}
   end
 
-  def handle_call({:get_pid, %Partition{} = partition}, _from, {sup, count} = state) do
+  def handle_call({:get_pid, %Partition{} = partition} = _message_passed_to_process, _from, {sup, count} = state) do
```
Comment (CJPoll) [inline lib/rivulet/sql_sink/writer/manager.ex]: `@spec` with `@doc`?

File: test/elastic_search_sink_test.exs [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+        unique_constraints: [["a", "b"], ["a"]],
+        primary_keys: ["c"],
+        table_pattern: name_pattern,
+        topic: topic
+      }
+
+    {:ok, %{config: config}}
+  end
+
+  describe "" do
```
Comment (CJPoll) [inline test/elastic_search_sink_test.exs]: ^

File: lib/rivulet/elastic_search_sink/database/elastic_search_generator.ex [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+    create_mapping(
+      config.elastic_url,
+      config.elastic_index,
+      config.elastic_type,
+      config.elastic_mapping
+    )
+  end
+  @spec create_mapping(url, index_name, type_name, mapping) :: true | exception
+  defp create_mapping(url, index, type, mapping) do
+    Logger.info "[Sinks.Elastic] creating elasticsearch mapping"
```
Comment (dconger) [inline lib/rivulet/elastic_search_sink/database/elastic_search_generator.ex]: So when you try to update an already existing field to a different type you get back this error:
screen shot 2018-04-09 at 1 13 06 pm

And we are handling that by raising which I actually think is the correct behavior in this case. Thoughts?

**Pull request merged**